In [21]:
import pandas as pd
df=pd.read_excel('Online retail.xlsx')
df

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [22]:
df.shape

(7500, 1)

In [23]:
df.columns

Index(['shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil'], dtype='object')

In [24]:
df.describe()

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
count                                                7500                                                                                                                                                                             
unique                                               5175                                                                                                                                                                             
top                                               cookies                                                                                                                                                                             
freq                                                  223

In [25]:
# Convert the dataframe into a list of lists for TransactionEncoder
# Each row in the dataframe is a transaction, and items are comma-separated
transactions_list = df.iloc[:,0].str.split(',').tolist()

In [26]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

te = TransactionEncoder()
te_array = te.fit(transactions_list).transform(transactions_list)
dataset = pd.DataFrame(te_array, columns=te.columns_)

In [27]:
transactions_list = df.iloc[:, 0].str.split(',').tolist()
te = TransactionEncoder()
te_array = te.fit(transactions_list).transform(transactions_list)
dataset = pd.DataFrame(te_array, columns=te.columns_)

In [36]:
frequent_itemsets = apriori(dataset, min_support=0.01, use_colnames=True)

In [37]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.2)
display(rules)

antecedents      consequents  antecedent support  \
0                     (avocado)  (mineral water)            0.033200   
1                     (burgers)           (eggs)            0.087200   
2                     (burgers)   (french fries)            0.087200   
3                     (burgers)      (green tea)            0.087200   
4                     (burgers)           (milk)            0.087200   
..                          ...              ...                 ...   
157           (spaghetti, milk)  (mineral water)            0.035467   
158  (mineral water, olive oil)      (spaghetti)            0.027467   
159      (spaghetti, olive oil)  (mineral water)            0.022933   
160   (mineral water, pancakes)      (spaghetti)            0.033733   
161       (spaghetti, pancakes)  (mineral water)            0.025200   

     consequent support   support  confidence      lift  representativity  \
0              0.238267  0.011467    0.345382  1.449559               1.0   
1              0.179733  0.028800    0.330275  1.837585               1.0   
2              0.170933  0.022000    0.252294  1.475976               1.0   
3              0.132000  0.017467    0.200306  1.517468               1.0   
4              0.129600  0.017867    0.204893  1.580964               1.0   
..                  ...       ...         ...       ...               ...   
157            0.238267  0.015733    0.443609  1.861817               1.0   
158            0.174133  0.010267    0.373786  2.146553               1.0   
159            0.238267  0.010267    0.447674  1.878880               1.0   
160            0.174133  0.011467    0.339921  1.952073               1.0   
161            0.238267  0.011467    0.455026  1.909736               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0    0.003556    1.163629       0.320785  0.044103   0.140620    0.196753  
1    0.013127    1.224782       0.499351  0.120941   0.183528    0.245256  
2    0.007095    1.108813       0.353289  0.093168   0.098135    0.190499  
3    0.005956    1.085415       0.373584  0.086583   0.078693    0.166315  
4    0.006566    1.094695       0.402580  0.089812   0.086504    0.171377  
..        ...         ...            ...       ...        ...         ...  
157  0.007283    1.369061       0.479911  0.060982   0.269572    0.254821  
158  0.005484    1.318826       0.549222  0.053659   0.241750    0.216373  
159  0.004802    1.379138       0.478747  0.040914   0.274910    0.245382  
160  0.005593    1.251163       0.504751  0.058384   0.200744    0.202885  
161  0.005462    1.397744       0.488682  0.045503   0.284561    0.251576  

[162 rows x 14 columns]

In [40]:
# Sort rules by lift in descending order to see the strongest positive associations first
high_lift_rules = rules.sort_values(by='lift', ascending=False)
display(high_lift_rules.head(10))

# Filter rules with a higher confidence threshold to identify more reliable associations
high_confidence_rules = rules[rules['confidence'] > 0.5].sort_values(by='confidence', ascending=False)
display(high_confidence_rules)

antecedents          consequents  \
75                      (herb & pepper)        (ground beef)   
152          (mineral water, spaghetti)        (ground beef)   
69                           (tomatoes)  (frozen vegetables)   
67                             (shrimp)  (frozen vegetables)   
145               (mineral water, milk)  (frozen vegetables)   
153        (mineral water, ground beef)          (spaghetti)   
143  (frozen vegetables, mineral water)               (milk)   
150               (mineral water, milk)        (ground beef)   
90                               (soup)               (milk)   
79                          (spaghetti)        (ground beef)   

     antecedent support  consequent support   support  confidence      lift  \
75             0.049467            0.098267  0.016000    0.323450  3.291555   
152            0.059733            0.098267  0.017067    0.285714  2.907540   
69             0.068400            0.095333  0.016133    0.235867  2.474134   
67             0.071333            0.095333  0.016667    0.233645  2.450820   
145            0.048000            0.095333  0.011067    0.230556  2.418415   
153            0.040933            0.174133  0.017067    0.416938  2.394361   
143            0.035733            0.129600  0.011067    0.309701  2.389672   
150            0.048000            0.098267  0.011067    0.230556  2.346223   
90             0.050533            0.129600  0.015200    0.300792  2.320923   
79             0.174133            0.098267  0.039200    0.225115  2.290857   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
75                1.0  0.011139    1.332841       0.732423  0.121457   
152               1.0  0.011197    1.262427       0.697745  0.121097   
69                1.0  0.009613    1.183913       0.639564  0.109304   
67                1.0  0.009866    1.180480       0.637444  0.111111   
145               1.0  0.006491    1.175740       0.616078  0.083669   
153               1.0  0.009939    1.416431       0.607207  0.086195   
143               1.0  0.006436    1.260904       0.603083  0.071737   
150               1.0  0.006350    1.171928       0.602713  0.081854   
90                1.0  0.008651    1.244836       0.599428  0.092158   
79                1.0  0.022088    1.163699       0.682292  0.168096   

     certainty  kulczynski  
75    0.249723    0.243136  
152   0.207875    0.229696  
69    0.155344    0.202549  
67    0.152887    0.204235  
145   0.149472    0.173320  
153   0.294000    0.257474  
143   0.206918    0.197546  
150   0.146705    0.171587  
90    0.196681    0.209038  
79    0.140672    0.312015

antecedents      consequents  antecedent support  \
134  (ground beef, eggs)  (mineral water)               0.020   
151  (ground beef, milk)  (mineral water)               0.022   

     consequent support   support  confidence      lift  representativity  \
134            0.238267  0.010133    0.506667  2.126469               1.0   
151            0.238267  0.011067    0.503030  2.111207               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
134  0.005368    1.544054       0.540548  0.040838   0.352354    0.274598  
151  0.005825    1.532756       0.538177  0.044409   0.347580    0.274738

#Based on the metrics:

#Rule Interpretations
1. (herb & pepper) → (ground beef):
    - Lift=3.29 suggests customers buying herb/pepper are more likely to buy ground beef.
2. (mineral water, spaghetti) → (ground beef):
    - Lift=2.9 indicates some association.
3. (tomatoes) → (frozen vegetables):
    - Lift=2.47 shows some positive relationship.
     and so on

#Implications
- These associations might guide product placement or promotions.From the generated rules, we observed patterns in customer purchasing behavior. Rules with higher lift values are particularly interesting as they indicate a strong positive correlation between items, suggesting that these items are purchased together more often than would be expected by chance. For example, we saw rules with relatively high lift involving combinations like (herb & pepper) and (ground beef).

- Rules with higher confidence indicate a higher probability that the consequent item(s) will be purchased when the antecedent item(s) are in the transaction. By filtering for higher confidence, we identified rules that are more reliable predictors of co-purchasing.

- In conclusion, association rule mining, using metrics like support, confidence, and lift, has allowed us to uncover interesting relationships between products in the online retail dataset. These insights can be valuable for business strategies such as optimizing product placement, creating effective product bundles or promotions, and improving targeted marketing efforts to enhance the customer shopping experience and potentially increase sales.


1.  **Support, Confidence, and Lift in Association Rule Mining:**
    -   **Support:** Measures how frequently an itemset appears in the dataset. It's the proportion of transactions containing the itemset.
    -   **Confidence:** Indicates the reliability of a rule X -> Y, showing how likely Y is purchased when X is purchased. It's the proportion of transactions containing X that also contain Y.
    -   **Lift:** Measures how much more likely Y is purchased when X is purchased compared to the probability of purchasing Y independently. A lift > 1 indicates a positive association, < 1 a negative association, and = 1 no association.

2.  **Limitations of Association Rule Mining:**
    *   High computational cost for large datasets.
    *   Generation of numerous trivial or non-actionable (spurious) rules.
    *   Difficulty handling continuous data (requires discretization).
    *   Challenges in finding associations for rare items due to minimum support thresholds.
    *   Interpretability issues with rules involving many items.
    *   Standard algorithms often ignore item quantities and purchase order.

### Data Analysis Key Findings

*   The transactional data was successfully transformed into a one-hot encoded format using `TransactionEncoder`, resulting in a DataFrame suitable for the Apriori algorithm.
*   Applying the Apriori algorithm with a minimum support of 0.01 identified frequent itemsets.
*   Association rules were generated from the frequent itemsets, filtered to include rules with a minimum confidence of 0.5.
*   Sorting the rules by 'lift' revealed strong positive associations between certain item combinations (e.g., `(ground beef, eggs)` -> `(mineral water)` with a confidence of 0.503).
*   An attempt to filter rules with higher confidence (> 0.6) and support (> 0.02) resulted in no rules meeting these stricter criteria simultaneously in this dataset.

### Insights or Next Steps

*   Focus on the rules with the highest lift values (> 1) as they represent the most interesting and potentially actionable associations for business strategies like product placement or bundling.
*   Explore adjusting the minimum support and confidence thresholds to find a balance between the number of generated rules and their significance, potentially considering lower thresholds if the current ones are too restrictive.


#Interview Questions

1.What is lift and why is it important  in association rules?
- Lift: Measures how much more likely items are bought together vs. independently.
- Importance: Lift > 1 indicates items are positively associated.

2.What is support and confidence and how do we calculate them?

- Support measures how frequently an itemset appears in transactions.
- Support = (Number of transactions with itemset) / (Total number of transactions).


3.What are some limitations or challenges of Association rules mining?
- 1. Setting thresholds: Support and confidence need careful tuning for meaningful rules.
- 2. Many rules generated: Can be overwhelming, needing filtering for relevance.
- 3. Interpretation requires context: Rules need business understanding for actionable insights.
- 4. Sparse data issues: Low support items may be missed or uninteresting rules.
